scripts/repartition_into_tracts.py

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env python
+
+"""
+Repartition a parquet file into tracts
+Author: Yao-Yuan Mao
+"""
+import os
+import multiprocessing as mp
+from argparse import ArgumentParser, RawTextHelpFormatter
+
+import numpy as np
+import pandas as pd
+from tqdm import tqdm
+
+import lsst.geom
+from lsst.daf.persistence import Butler
+import desc_dc2_dm_data
+
+
+__all__ = ["get_tract_patch", "repartition_into_tracts"]
+
+
+def get_tract_patch(skymap, ra, dec):
+    radec = lsst.geom.SpherePoint(ra, dec, lsst.geom.degrees)
+    tractInfo = skymap.findTract(radec)
+    patchInfo = tractInfo.findPatch(radec)
+    return tractInfo.getId(), "{},{}".format(*patchInfo.getIndex())
+
+
+def get_tract_patch_arrays(skymap, ra_arr, dec_arr, disable_tqdm=None):
+    tract, patch = [], []
+    for ra, dec in tqdm(zip(ra_arr, dec_arr), total=len(ra_arr), disable=disable_tqdm):
+        tract_this, patch_this = get_tract_patch(skymap, ra, dec)
+        tract.append(tract_this)
+        patch.append(patch_this)
+    return tract, patch
+
+
+def repartition_into_tracts(
+    input_file,
+    output_root_dir,
+    skymap_source_repo,
+    ra_label="ra",
+    dec_label="dec",
+    n_cores=None,
+    silent=False,
+    **kwargs
+):
+    """ Take a parquet catalog and split it into tracts according to a given skymap, and write to disk
+
+    Parameters
+    ----------
+    input_file : str
+        Path to the input parquet file.
+    output_root_dir : str
+        Path to the output directory. The output files will have the following filename
+        <output_root_dir>/<tract>/<input_file_basename>
+    skymap_source_repo : str
+        Path or existing key if desc_dc2_dm_data.REPOS to indicate the bulter repo for loading skymap
+
+    Optional Parameters
+    ----------------
+    ra_label : str, optional
+        Column name for RA, default to 'ra'. The unit is assumed to be degrees.
+    dec_label : str, optional
+        Column name for Dec, default to 'dec'. The unit is assumed to be degrees.
+    silent : bool, optional (default: False)
+        If true, turn off most printout.
+    """
+    my_print = (lambda x: None) if silent else print
+    tqdm_disable = silent or None
+
+    repo = desc_dc2_dm_data.REPOS.get(skymap_source_repo, skymap_source_repo)
+    my_print("Obtain skymap from", repo)
+    skymap = Butler(repo).get("deepCoadd_skyMap")
+
+    my_print("Loading input parquet file", input_file)
+    df = pd.read_parquet(input_file)
+
+    n_cores = n_cores or os.cpu_count() or 1
```

Comment:
Thanks, that's good idea. One caveat is that `os.sched_getaffinity` only available on certain unix platforms, but that should be fine for our use. To be careful, I added some try/except just in case. A new function called `get_number_of_workers` to improve readability.